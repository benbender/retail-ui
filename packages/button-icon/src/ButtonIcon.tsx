import { Spinner } from '@retail-ui/spinner'
import { Theme, useThemeCtx } from '@retail-ui/theme'
import { cloneElement } from '@retail-ui/utils'
import clsx from 'clsx'
import React, { ButtonHTMLAttributes } from 'react'

type Ref = HTMLButtonElement
type ReactButtonProps = Omit<
  ButtonHTMLAttributes<HTMLButtonElement>,
  'disabled'
>

type ButtonIconColor = keyof Theme['ButtonIconStyles']['variant']['default']
type ButtonIconVariant = keyof Theme['ButtonIconStyles']['variant']
type ButtonIconSize = keyof Theme['ButtonIconStyles']['size']
type ButtonIconShape = keyof Theme['ButtonIconStyles']['shape']

export type ButtonIconProps = {
  size?: ButtonIconSize
  variant?: ButtonIconVariant
  shape?: ButtonIconShape
  color?: ButtonIconColor
  spinnerClassName?: string
  icon: React.ReactNode
  isLoading?: boolean
  isDisabled?: boolean
}

export const ButtonIcon = React.forwardRef<
  Ref,
  ReactButtonProps & ButtonIconProps
>((props, ref) => {
  const {
    className,
    spinnerClassName,
    isDisabled,
    isLoading,
    variant = 'default',
    color = 'primary',
    size = 'base',
    icon,
    shape,
    onClick,
    ...rest
  } = props

  const {
    theme: { ButtonIconStyles },
  } = useThemeCtx()

  const variantCls = ButtonIconStyles.variant

  const colorCls = variantCls[variant]
  const sizeCls = ButtonIconStyles.size
  const shapeCls = ButtonIconStyles.shape

  const cls = clsx(
    className,
    ButtonIconStyles.base,
    isDisabled && ButtonIconStyles.disabled,
    isLoading && ButtonIconStyles.loading,
    colorCls[color],
    sizeCls[size],
    shape && shapeCls[shape],
  )

  const spinnerSizeCls = ButtonIconStyles.spinner.size
  const spinnerCls = clsx('absolute', spinnerSizeCls[size], spinnerClassName)
  const isLoadingCls = isLoading ? 'opacity-0' : 'opacity-100'

  /* Icons */
  const iconSizeCls = ButtonIconStyles.icon.size
  const iconCls = clsx(iconSizeCls[size], isLoadingCls)

  const iconNode = cloneElement(icon, { className: iconCls })

  return (
    <button
      ref={ref}
      className={cls}
      disabled={isDisabled}
      onClick={onClick}
      {...rest}
    >
      {isLoading && <Spinner className={spinnerCls} />}
      {iconNode}
    </button>
  )
})

ButtonIcon.displayName = 'ButtonIcon'
