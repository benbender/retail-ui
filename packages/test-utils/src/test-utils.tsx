import { defaultTheme, ThemeProvider } from '@retail-ui/theme'
import {
  render as RtlRender,
  RenderOptions,
  RenderResult,
} from '@testing-library/react'
import * as React from 'react'

type UI = Parameters<typeof RtlRender>[0]

// UI-less passthrough fallback to prevent using conditional logic in render
function ChildrenPassthrough({ children }: { children: React.ReactElement }) {
  return children
}

export interface TestOptions extends Omit<RenderOptions, 'wrapper'> {
  /**
   * optional additional wrapper, e.g. Context
   *
   * @example
   * ```ts
   * // single wrapper
   * render(<MyConponent />, {
   *  wrapper: MyContext
   * });
   *
   * // multiple wrapper
   * render(<MyConponent />, {
   *  wrapper: ({ children }) => (
   *    <ContextA>
   *      <ContextB>
   *        {children}
   *      <ContextB />
   *    <ContextA />
   *  )
   * });
   *
   * ```
   */
  wrapper?: typeof ChildrenPassthrough
}

/**
 * Custom render for @testing-library/react
 *
 * @see https://testing-library.com/docs/react-testing-library/setup#custom-render
 * @param component the component under test
 * @param options customized test options
 */
export const render = (
  ui: UI,
  { wrapper: Wrapper = ChildrenPassthrough, ...options }: TestOptions = {},
): RenderResult => {
  return RtlRender(
    <ThemeProvider value={{ theme: defaultTheme }}>
      <Wrapper>{ui}</Wrapper>
    </ThemeProvider>,
    options,
  )
}

export { fireEvent } from '@testing-library/react'
