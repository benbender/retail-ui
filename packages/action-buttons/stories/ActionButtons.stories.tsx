import * as React from 'react'

import { ActionButtons, ActionButtonsProps } from '../src'

const meta = {
  title: 'Retail-UI/ActionButtons',
  component: ActionButtons,
}

export default meta

const Template = (args: ActionButtonsProps) => (
  <ActionButtons {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {}
