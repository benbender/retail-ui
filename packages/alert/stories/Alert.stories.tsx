import * as React from 'react'

import { Alert, AlertProps } from '../src'

const meta = {
  title: 'Retail-UI/Alert',
  component: Alert,
}

export default meta

const Template = (args: AlertProps) => (
  <Alert {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'A alert that informs you of stuff',
}
