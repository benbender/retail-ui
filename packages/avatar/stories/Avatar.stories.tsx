import * as React from 'react'

import { Avatar, AvatarProps } from '../src'

const meta = {
  title: 'Retail-UI/Avatar',
  component: Avatar,
}

export default meta

const Template = (args: AvatarProps) => (
  <Avatar {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  src: `https://pbs.twimg.com/profile_images/1285102166709919744/XvlWrJny_400x400.jpg`,
}
