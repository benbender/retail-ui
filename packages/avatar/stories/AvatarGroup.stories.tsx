import * as React from 'react'

import { Avatar, AvatarGroup, AvatarProps } from '../src'

const meta = {
  title: 'Retail-UI/AvatarGroup',
  component: AvatarGroup,
}

export default meta

const src = `https://pbs.twimg.com/profile_images/1285102166709919744/XvlWrJny_400x400.jpg`

const Group = (args: AvatarProps) => (
  <AvatarGroup data-testid={meta.title}>
    <Avatar src={src} {...args} />
    <Avatar title="PJ" size="base" {...args} />
    {/* <Avatar icon={UserSolid} {...args} /> */}
  </AvatarGroup>
)

export const Simple = Group.bind({})
// @ts-ignore
Simple.args = {}
