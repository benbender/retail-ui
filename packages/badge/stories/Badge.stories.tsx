import * as React from 'react'

import { Badge, BadgeProps } from '../src'

const meta = {
  title: 'Retail-UI/Badge',
  component: Badge,
}

export default meta

const Template = (args: BadgeProps) => (
  <Badge {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'Badge',
}
