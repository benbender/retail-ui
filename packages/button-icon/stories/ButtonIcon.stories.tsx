import { AcademicCapSolid } from '@retail-ui/heroicons'
import * as React from 'react'

import { ButtonIcon, ButtonIconProps } from '../src'

const meta = {
  title: 'Retail-UI/ButtonIcon',
  component: ButtonIcon,
}

export default meta

const Template = (args: ButtonIconProps) => (
  <ButtonIcon icon={<AcademicCapSolid />} {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {}
