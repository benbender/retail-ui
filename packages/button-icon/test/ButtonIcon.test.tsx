import { AcademicCapSolid } from '@retail-ui/heroicons'
import { render } from '@retail-ui/test-utils'
import * as React from 'react'

import * as ButtonIcon from '../stories/ButtonIcon.stories'

describe('ButtonIcon', () => {
  it('renders correctly', () => {
    const { asFragment } = render(
      <ButtonIcon.Simple icon={<AcademicCapSolid />} />,
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
