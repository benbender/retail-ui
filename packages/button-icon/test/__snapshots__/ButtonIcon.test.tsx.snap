// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonIcon renders correctly 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center leading-5 align-bottom transition-colors duration-150 ease-in transform scale-100 border cursor-pointer focus:outline-none dark:focus:shadow-outline-gray active:scale-95 overflow-hidden no-underline select-none text-white bg-purple-600 border-transparent active:shadow-outline-purple focus:bg-purple-700 hover:bg-purple-700 p-2 rounded-lg"
    data-testid="Retail-UI/ButtonIcon"
  >
    <svg
      aria-hidden="true"
      class="w-7 h-7 text-current fill-current !h-4 !w-4 opacity-100"
      fill="none"
      viewBox="0 0 20 20"
    >
      <path
        d="M10.394 2.08a1 1 0 00-.788 0l-7 3a1 1 0 000 1.84L5.25 8.051a.999.999 0 01.356-.257l4-1.714a1 1 0 11.788 1.838L7.667 9.088l1.94.831a1 1 0 00.787 0l7-3a1 1 0 000-1.838l-7-3zM3.31 9.397L5 10.12v4.102a8.969 8.969 0 00-1.05-.174 1 1 0 01-.89-.89 11.115 11.115 0 01.25-3.762zM9.3 16.573A9.026 9.026 0 007 14.935v-3.957l1.818.78a3 3 0 002.364 0l5.508-2.361a11.026 11.026 0 01.25 3.762 1 1 0 01-.89.89 8.968 8.968 0 00-5.35 2.524 1 1 0 01-1.4 0zM6 18a1 1 0 001-1v-2.065a8.935 8.935 0 00-2-.712V17a1 1 0 001 1z"
      />
    </svg>
  </button>
</DocumentFragment>
`;
